src/Formality__Form.re

```diff
@@ -143,65 +156,51 @@ module Make = (Form: Form) => {
         switch (validator) {
         | None => React.Update({...state, data})
         | Some(validator) =>
-          let emitted = state.emittedFields->Set.has(field);
-          switch (validator.strategy, emitted, state.submittedOnce) {
-          | (_, true, _)
-          | (_, _, true)
-          | (Strategy.OnFirstChange, false, false) =>
+          switch (validator.strategy, state.submittedOnce) {
+          | (_, true)
+          | (Strategy.OnFirstChange, false) =>
             let result = data->(validator.validate);
-            let (results, emittedFields) =
+            let results =
               validator.dependents
               ->validateDependents(
                   ~data,
                   ~validators=state.validators,
                   ~results=state.results,
-                  ~emittedFields=state.emittedFields,
                 );
             React.Update({
               ...state,
               data,
-              results: results->Map.set(field, result),
-              emittedFields: emittedFields->Set.add(field),
+              results: results->Map.set(field, Dirty(result)),
             });
 
           | (
               Strategy.OnFirstSuccess | Strategy.OnFirstSuccessOrFirstBlur,
               false,
-              false,
             ) =>
             let result = data->(validator.validate);
-            let (results, emittedFields) =
+            let results =
               validator.dependents
               ->validateDependents(
                   ~data,
                   ~validators=state.validators,
                   ~results=state.results,
-                  ~emittedFields=state.emittedFields,
                 );
-            switch (result) {
-            | Ok(Valid | NoValue) =>
-              React.Update({
-                ...state,
-                data,
-                results: results->Map.set(field, result),
-                emittedFields: emittedFields->Set.add(field),
-              })
-            | Error(_) =>
-              React.Update({...state, data, results, emittedFields})
-            };
+            React.Update({
+              ...state,
+              data,
+              results: results->Map.set(field, Dirty(result)),
+            });
```

Comment:
One more FYI comment: this `switch` shouldn't be removed b/c otherwise it turns `Strategy.OnFirstSuccess | Strategy.OnFirstSuccessOrFirstBlur` into `Strategy.OnFirstChange`. But w/ proposed changes it will be handled differently.